src/metabase/api/email.clj

```diff
@@ -77,15 +77,27 @@
              (name (mb-to-smtp-settings k))
              (str/upper-case v))])))
 
+(defn- env-var-values-by-email-setting
+  "Returns a map of setting names (keywords) and env var values.
+   If an env var is not set, the setting is not included in the result."
+  []
+  (into {}
+        (for [setting-name (keys mb-to-smtp-settings)
+              :let         [value (setting/env-var-value setting-name)]
+              :when        (some? value)]
+          [setting-name value])))
+
 (api/defendpoint PUT "/"
   "Update multiple email Settings. You must be a superuser or have `setting` permission to do this."
   [:as {settings :body}]
   {settings su/Map}
   (validation/check-has-application-permission :setting)
-  ;; the frontend has access to an obfuscated version of the password. Watch for whether it sent us a new password or
-  ;; the obfuscated version
-  (let [obfuscated? (and (:email-smtp-password settings) (email/email-smtp-password)
+  (let [;; the frontend has access to an obfuscated version of the password. Watch for whether it sent us a new password or
+        ;; the obfuscated version
+        obfuscated? (and (:email-smtp-password settings) (email/email-smtp-password)
                          (= (:email-smtp-password settings) (setting/obfuscate-value (email/email-smtp-password))))
+        ;; override `nil` values in the request with environment variables
+        settings    (merge settings (env-var-values-by-email-setting))
```

Comment:
When a setting is set via env variable, the env variable value is returned automatically when you fetch the value of the setting, taking precedence over any value in the DB. It shouldn't be necessary to re-write these values back into the setting. In fact I think it could lead to incorrect behavior, because if the instance is restarted without the env variables, the values would still be in the DB for these settings but the user wouldn't expect that.

Could you change this to include the env variable values when testing the connection, but not when saving the new settings?

Reply:
Good catch, I hadn't considered that.